### PR TITLE
i#4194: Adds dr_merge_arith_flags()

### PR DESCRIPTION
**Pull request by johnfxgalea**
Adds a convenience routine, namely dr_merge_arith_flags(), to merge arithmetic flags during restoration done by outlined code.

Fixes: #4194

### CODE REVIEW

File: core/lib/instrument.c [hunk 1]
```diff
@@ -6064,6 +6064,24 @@ dr_restore_arith_flags_from_reg(void *drcontext, instrlist_t *ilist, instr_t *wh
...
 
+DR_API reg_t
+dr_merge_arith_flags(reg_t cur_xflags, reg_t saved_xflag)
+{
+#    ifdef AARCHXX
+    cur_xflags &= ~(EFLAGS_ARITH);
+    cur_xflags |= saved_xflag;
+#    elif defined(X86)
+    uint sahf;
+    sahf = (saved_xflag & 0xff00) >> 8;
```
Comment (derekbruening) [inline core/lib/instrument.c]: nit: can combine decl and assign

File: core/lib/instrument_api.h [hunk 2]
```diff
@@ -5128,8 +5128,15 @@ void
 dr_restore_arith_flags_from_reg(void *drcontext, instrlist_t *ilist, instr_t *where,
                                 reg_id_t reg);
 
-/* FIXME PR 315333: add routine that scans ahead to see if need to save eflags.  See
- * forward_eflags_analysis(). */
+DR_API
+/**
+ * A convenience routine to aid restore the arith flags done by outlined code,
```
Comment (derekbruening) [inline core/lib/instrument_api.h]: s/restore/restoring/

File: ext/drbbdup/drbbdup.c [hunk 3]
```diff
@@ -1122,15 +1122,8 @@ drbbdup_prepare_redirect(dr_mcontext_t *mcontext, drbbdup_manager_t *manager,
...
-        reg_t newval = mcontext->xflags;
-        val = (reg_t)drbbdup_get_tls_raw_slot_val(DRBBDUP_FLAG_REG_SLOT);
-        sahf = (val & 0xff00) >> 8;
-        newval &= ~(EFLAGS_ARITH);
-        newval |= sahf;
-        if (TEST(1, val)) /* seto */
-            newval |= EFLAGS_OF;
+        reg_t val = (reg_t)drbbdup_get_tls_raw_slot_val(DRBBDUP_FLAG_REG_SLOT);
+        reg_t newval = dr_merge_arith_flags(mcontext->xflags, val);
         mcontext->xflags = newval;
```
Comment (derekbruening) [inline ext/drbbdup/drbbdup.c]: Could just assign directly and eliminate the `newval` var.

File: api/docs/release.dox [hunk 4]
```diff
@@ -250,6 +250,8 @@ Further non-compatibility-affecting changes include:
  - Added drwrap_get_stats().
  - Added #DRWRAP_NO_DYNAMIC_RETADDRS for reducing drwrap overhead at the cost
    of missing some post-call callbacks.
+ - Added dr_merge_arith_flags() as a convenience routine to merge arithmetic flags
+   for restoration done by outline code.
```
Comment (derekbruening) [inline api/docs/release.dox]: s/outline/outlined/?  I don't think "outline code" is going to be familiar to many people; "outlined" makes it clearer IMHO.

File: core/lib/instrument.c (hunk 1 above)
Comment (johnfxgalea) [reply core/lib/instrument.c]: Done.

File: core/lib/instrument_api.h (hunk 2 above)
Comment (johnfxgalea) [reply core/lib/instrument_api.h]: Done/

File: api/docs/release.dox (hunk 4 above)
Comment (johnfxgalea) [reply api/docs/release.dox]: Done.

**Pull request merged**